### ISSUE TITLE
Color blending across layers

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#79]
## ❓ Questions and Help

A use case that I think will come up frequently is having multiple color channels of data that we want to visualize simultaneously (probably most commonly between 2-6 at a time). I think we want to create an independent image layer for each color channel, assign each it's own colormap with independently adjustable clims (which we can do now), but what we then lack is the ability to do any form of color blending (beyond a global opacity). FIJI does this quite nicely - see bellow two channels of information being blended together (one green / one magenta). Does anyone have any thoughts on how to implement this behaviour using vispy? @royerloic @freeman-lab - I tried digging around a bit to see if blending modes were present and I couldn't find any. I can imaging how we would do these computations in the cpu using numpy (but I'm guessing that's not a popular option!!!).

Also as side note vispy lacks a lot of the colormaps that we'll want for cell biology that go from black to a particular color (like red, green, blue, cyan, yellow, magenta), and I think we'll need to add them.

![image](https://user-images.githubusercontent.com/6531703/49310763-0de13400-f494-11e8-9916-9a0b489a9780.png)

**Comment by jni** [comment]
In a related note, we should also offer alpha proportional to the image intensity, that can be a quite nice way to overlay two images. Could be a tick box next to the alpha slider.

I'll leave the blending modes to the graphics experts. Yes, I myself would solve it with NumPy, and @royerloic would yell at me. =P

**Comment by sofroniewn** [comment]
Alpha proportional to intensity sounds good. I get how to do it using numpy, though I'd have to first convert the image to RGBA, which will require looking at the clims / making some decisions about how to map to 0-1 ...

**Comment by sofroniewn** [comment]
So looking into vispy a little, a couple things to note. First it's very easy to create new colormaps in vispy, including ones with alpha proportional to intensity.
```python
blue = Colormap([(0, 0, 0, 1), (0., 0., 1., 1.)])
blue_alpha = Colormap([(0, 0, 0, 0), (0., 0., 1., 1.)])
```
The first colormap is one that goes from black to blue at full opacity, the second one does the same, but now alpha is proportional to intensity too.

In general, I think we should create a curated list of colormaps likely to be used in science and make sure they can be selected by the drop down menu, and maybe remove some of the more esoteric vispy defaults. Once this is done, for many colormaps it will then be easy to have an alpha proportional to intensity version and a normal version.

However, the alpha proportional version still doesn't yield very nice results. It's definitely better than nothing, but it's not the same as color blending. The order of the layers still matters, and in general the images have a sort of grey, washed out look to them. See below of a black to blue and black to red image with intensity proportional to alpha in both.

![image](https://user-images.githubusercontent.com/6531703/49333778-b4533500-f579-11e8-8b29-e01b3c9a7349.png)

Further advice / comments appreciated. I'll keep poking around too

**Comment by sofroniewn** [comment]
On the left a real rgb image (happened to be 3 channels worth of data) on right 3 channels plotted seperately with colormaps of the form `Colormap([(0, 0, 0, 0), (0., 0., 1., 1.)])` for the 3 different channels (r, g, b) adjusted appropriately. I think we should strive to be able to reproduce images that look like the ones on the left from data that has been separately plotted, when in the 2-6 channel regime.
![image](https://user-images.githubusercontent.com/6531703/49333871-adc5bd00-f57b-11e8-8d30-83d758006a62.png)

**Comment by freeman-lab** [comment]
👍 to using the blending of that RGB image (or similar) as the test case for what we want to achieve

👍 to supporting alpha, even if it doesn’t do everything blending modes can

👍 to curating a list of colormaps -- awesome it's so easy to make custom ones in vispy. i'd be careful to avoid overindexing on the biological use cases per se, but i'm sure we could come up with a relatively manageable list of "good" colormaps that included those ones, most of the currently in vogue perceptually uniform colormaps, and a few other common ones.

😞 if vispy doesn't support blending modes out of the box, maybe worth reaching out to the developers to see if we're missing something? just open an issue? would be pretty complex to try and do that well ourselves (whether in the CPU or in the GPU via custom shaders)

**Comment by sofroniewn** [comment]
ok nice, I'll try creating a minimal vispy example that doesn't use napari but sets up the visualization in the same way. Once I have that working I'll think about posting a vispy issue!

**Comment by jni** [comment]
Yes, I wasn't suggesting proportional alpha as a blending solution, rather, I've used it for example to overlay light microscopy and EM in CLEM datasets, it works really nicely there. I can't remember whether I used a linear mapping actually, maybe it was more exaggerated.

This paper has some nice cyclic colourmaps, which I don't think is true of matplotlib currently, mainly because they constrained themselves to be colourblind friendly, which I think is *very* hard in cyclic space and overly restrictive:
http://arxiv.org/abs/1509.03700

Finally, on the converse side, I've seen people complain that the standard RGB colour channels used in most microscopy and recommend CMY as the default colour blending. I must say I'm a fan of the look!
https://twitter.com/calebadawson/status/1052107086367092736

It would be cool to have this be the default in Napari, though I can see some people being surprised by this, so I wouldn't push hard for it.

**Comment by sofroniewn** [comment]
That's a great thread - I'm open to supporting CMY as a default, with RGB as an option, but right now we havn't got either working!!

**Comment by royerloic** [comment]
Guys, I checked vispy. We can do this by tweaking the OpenGL options at the canvas level by disabling the depth testing and enabling blending. Let me take care of this...  It's exactly what we did for supporting multi-channels in ClearVolume and its the right thing to do to have it work for 1D, 2D, and 3D.

**Comment by sofroniewn** [comment]
@royerloic sounds great, let me know when there's something you want me to look over

also here's a link to the behavior in imageJ https://imagej.net/Color_Image_Processing, I think we want to support functionality like their merge channel option.

**Comment by sofroniewn** [comment]
@royerloic - turns out this is pretty easy, thanks for turning me in the right direction though - it's implemented in vispy as `set_gl_state` on an `ImageVisual` (and other visuals). They have three presets `opaque`, `translucent`, and `additive`. Right now we're essentially using `translucent`. `additive` achieves the behavior I desired above. `opaque` doesn't allow for any blending and doesn't seem too useful, but might have a place.

I propose to add a property `_gl_state` to `_base_layer` and a dropdown menu to the layers widget with the 3 different gl_state presets. It will default to `translucent`, but now it will be easy to change it to the others.

If you'd like to see it in action, just create a viewer as normal and then do the following:
`viewer.layers[1]._node.set_gl_state('additive')`

**Comment by sofroniewn** [comment]
Merged in #81
